---
title: "Differential expression analysis figures Indoor vs Outdoor"
author: "Laura García Romañach"
date: '`r format(Sys.time(), "%d.%m.%Y")`'
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE)

library(knitr)
library(tidyverse)
library(ggrepel)
library(gplots)
library(pheatmap)
library(cowplot)
library(readxl)
library(RColorBrewer)
library(forcats)
library(extrafont)
```

```{r functions}
gg_color_hue <- function(n) {
    hues = seq(15, 375, length = n + 1)
    hcl(h = hues, l = 65, c = 100)[1:n]
  }

plot_GO <- function(data, ...) {
  ggplot(data %>% 
           filter(...), 
  aes(x = Regulation, y = `GO term`, 
                        color = `P-value`, size = x)) + 
  geom_point() +
  scale_color_continuous(low = "#E08214" , high = "#542788") +
   scale_size_continuous(range = c(1, 7)) +
  theme_bw() + 
  labs(x = "",
       y = "") + 
  labs(color = "p-value",
       size = "Number of DEGs") +
    theme(axis.text = element_text(size = 10, family = "Arial", color = "black"),
          axis.text.x = element_text(size = 12, color = "black"),
          axis.ticks = element_line(size = 0.1, color = "black"),
          line = element_line(size = 0.1),
          axis.line = element_line(size = 0.05, color = "black"),
          panel.border = element_rect(linewidth =  0.1, color = "black"),
          legend.key.size = unit(0.5, "cm"),
          legend.text = element_text(size = 6),
          legend.title = element_text(size = 8)) 
}

sel_genes_season <- function(season, ...){
  selected_season <- selected_genes %>% 
   filter(Season == season)
  
  genes_season <- filter(expression_data, Gene %in% selected_season$Gene) %>%
        group_by(Gene, `Month/Treatment`) %>%
        summarize(mean = mean(Expression)) %>% 
        filter(...)
  
  genes_season$`Month/Treatment` <- factor(genes_season$`Month/Treatment`, levels = c("SEP",
                                                                    "OCT",
                                                                    "DEC",
                                                                    "JAN",
                                                                    "FEB",
                                                                    "MAR",
                                                                    "APR",
                                                                    "MAY",
                                                                    "JUN",
                                                                    "JUL",
                                                                    "AUG",
                                                                    "SD15","CT2","CT8","CT10",
                                                                "LD1","LD2","LD3", "LD4",
                                                                "SD1","SD2","SD3","SD10"))
        genes_season <- genes_season %>%
          arrange(`Month/Treatment`) %>%
        pivot_wider(names_from =  `Month/Treatment`,
                    values_from = mean) %>%
        column_to_rownames("Gene") %>% 
        mutate(Total = rowSums(.)) %>% 
          filter(Total != 0) %>% 
          dplyr::select(!Total)
         
        custom_order <- as.vector(selected_season %>%
           dplyr::select(Gene))
        genes_season <- genes_season[match(custom_order$Gene, rownames(genes_season)), ]

}

get_heatmap <- function(samples_season, genes_season, file_name) {
  #dist.var <- as.dist(1-cor(data.loc))
  # dist.var <- dist(t(sel_genes_winter), method="euclidean")
  # dist.var.tree <- hclust(dist.var, method = "ward.D")
  # 
  # dist.obs <- as.dist(1-cor(t(sel_genes_winter)))
  # dist.obs.tree <- hclust(dist.obs, method = "ward.D")

  annot_col <- data.frame(`Month/Treatment` = samples_season$`Month/Treatment`,
   Location = samples_season$Location,
    check.names = FALSE) %>% 
    unique()

  annot_col$`Month/Treatment` <- factor(annot_col$`Month/Treatment`, levels = c("SEP",
                                                                                "OCT",
                                                                                "DEC",
                                                                                "JAN",
                                                                                "FEB",
                                                                                "MAR",
                                                                                "APR",
                                                                                "MAY",
                                                                                "JUN",
                                                                                "JUL",
                                                                                "AUG",
                                                                                "SD15","CT2","CT8","CT10",
                                                                                "LD1","LD2","LD3", "LD4",
                                                                         "SD1","SD2","SD3","SD10"))
  annot_col <- annot_col %>% 
    arrange(`Month/Treatment`)
  
    
  rownames(annot_col) <- colnames(genes_season)
  
  vals <- unique(samples.sep$Location)
  location_map <- gg_color_hue(length(vals))
  names(location_map) <- vals

  annot_colors = list(
    Location = location_map,
     `Month/Treatment` = c(
        SEP = "#F3766E",
        OCT = "#D98F28",
        DEC = "#2CBAA4",
        JAN = "#07BADE",
        FEB = "#499ED7",
        MAR = "#ADA334",
        APR = "#65B345",
        MAY = "#2CB45D",
        JUN = "#9D86BE",
        JUL = "#C574AF",
        AUG = "#F067A6",
        SD15 = "#D98F28",
        CT2 = "#2CBAA4",
        CT8 = "#07BADE",
        CT10 = "#499ED7",
        LD1 = "#ADA334",
        LD2 = "#65B345",
        LD3 = "#2CB45D",
        LD4 = "#346749",
        SD1 = "#9D86BE",
        SD2 = "#C574AF",
        SD3 = "#F067A6",
        SD10 = "#F3766E"
        
    )
  )
  
custom_breaks <- seq(-1.5, 1.5, length.out = 9)  

pdf(file = file_name, height = 7, width = 4)
pheatmap(mat = as.matrix(genes_season),
           cluster_rows = FALSE,
           cluster_cols = FALSE, #dist.var.tree.GLOBAL,
           scale = "row",
           legend = TRUE,
           border_color = NA,
           color = colorRampPalette(c("dodgerblue","white","firebrick"))(8),
         breaks = custom_breaks,
           fontsize = 8,
           fontsize_row = 7,
           fontsize_col = 10,
           fontfamily = "Arial",
           # srtCol = 45,
           show_rownames = TRUE,
           show_colnames = FALSE,
           #labels_col = names,
           annotation_legend = TRUE,
           annotation_col = annot_col,
           #annotation_row = annot_row,
           annotation_colors = annot_colors,
           name = "Expression")
dev.off()
  }
```

#Load data 
```{r load data GO}
selected_go <- read_excel("data/Selected_GO_terms_and_DEGs.xlsx", sheet = 1) 
selected_go$`P-value` <- as.numeric( selected_go$`P-value`)
```

```{r load expression data, message=FALSE, warning=FALSE}
load("data/gene-counts-and-VSTexpression.RData")
cat("Genes x Samples: ", paste0(dim(data), collapse = " x "))

samples.sep <- samples.sep %>% 
  unite("Treatment_Week", Treatment, Week, na.rm = TRUE, sep = "") 

samples.sep$Treatment_Week <- if_else(samples.sep$Location == "Indoor" , samples.sep$Treatment_Week, "")

samples.sep <- samples.sep %>% 
  unite("Month/Treatment", Month, Treatment_Week, na.rm = TRUE, sep = "")

treatment_mapping <- c(
  "SEP", "OCT", "DEC", "JAN", "FEB", "MAR", "APR", "MAY", "JUN", "JUL",
  "AUG", "SD15", "CT2", "CT8", "CT10", "LD1", "LD2", "LD3", "LD4", "SD1", 
  "SD2", "SD3", "SD10"
)

samples.sep$Treatment2 <- match(samples.sep$`Month/Treatment`, treatment_mapping)

samples.sep <- samples.sep %>%
  arrange(desc(Location), Tissue) %>% 
  arrange(Treatment2)

samples.sep$Location <- factor(samples.sep$Location,
                               levels = c("Outdoor", 
                                          "Indoor"))

samples.sep.winter <- samples.sep %>% 
    filter(`Month/Treatment` == "DEC" |
             `Month/Treatment` == "JAN" |
             `Month/Treatment` == "FEB" |
             `Month/Treatment` == "CT8" |
             `Month/Treatment` == "CT10") %>% 
  filter(!grepl("A", S2)) 

samples.sep.summer <- samples.sep %>% 
  filter(S1 == "JUN" | 
           S1 == "JUL" |
           `Month/Treatment` == "LD4") %>% 
  filter(!grepl("A", S2))

samples.sep.autumn <- samples.sep %>% 
  filter(S1 == "SEP" | 
          S1 == "OCT"| 
          S2 == "SDW15" |
            S1 == "SDW15" |
           S2 == "CTW2" | 
           S1 == "CTW2") %>% 
  filter(!grepl("A", S2))

expression_data <- data %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "Gene") %>% 
  pivot_longer(!Gene, names_to = "Samples", values_to = "Expression") %>% 
  mutate(Sample = factor(Samples, levels = colnames(data))) %>% 
  left_join(samples.sep %>% dplyr::select(-Expression), by = "Samples") %>% 
  arrange(Treatment2) %>% 
   filter(!grepl("A", S2))

selected_genes <- read_excel("data/Selected_GO_terms_and_DEGs.xlsx", sheet = 2) 
```

#Plot of selected GO terms
```{r plots}
#Order the dataset for plotting
selected_go$Regulation <- factor(selected_go$Regulation, levels = c("up", "down"),
                                   labels = c("Upregulated", "Downregulated"))
selected_go$Season <- factor(selected_go$Season, levels = c("summer", "autumn", "winter"))
selected_go <- dplyr::arrange(selected_go, Season, Regulation, x)
selected_go$`GO term`  <- factor(selected_go $`GO term` , levels = selected_go$`GO term`)
selected_go$`GO term` <- fct_rev(selected_go$`GO term`)

#Plots
go_up <- plot_GO(selected_go, Regulation == "Upregulated") + theme(legend.position = "none")
go_down <-  plot_GO(selected_go, Regulation == "Downregulated") +  scale_y_discrete(position = "right") 
legend <- get_legend(go_down)

plot_grid(go_up, go_down +
            theme(legend.position = "none"), legend, nrow = 1, rel_widths = c(1, 0.91, 0.25)) 

# ggsave("Figures/Transcriptomics_Torgeir/GO_all2.pdf", width = 12, height = 8)
```

#Heatmaps of selected genes
```{r heatmaps of selected genes}
sel_genes_winter <- sel_genes_season("winter", `Month/Treatment` == "DEC" |
                 `Month/Treatment` == "JAN" |
                 `Month/Treatment` == "FEB" |
                 `Month/Treatment` == "CT8" |
                 `Month/Treatment` == "CT10")

sel_genes_summer <- sel_genes_season("summer", `Month/Treatment` == "JUN" |
                 `Month/Treatment` == "JUL" |
                 `Month/Treatment` == "LD4")

sel_genes_autumn <- sel_genes_season("autumn", `Month/Treatment` == "SEP" |
                 `Month/Treatment` == "OCT" |
                 `Month/Treatment` == "SD15" |
                   `Month/Treatment` == "CT2")
  
# get_heatmap(samples.sep.winter, sel_genes_winter, "Figures/Transcriptomics_Torgeir/heatmap_selected_genes_winter.pdf")
# get_heatmap(samples.sep.autumn, sel_genes_autumn, "Figures/Transcriptomics_Torgeir/heatmap_selected_genes_autumn.pdf")
# get_heatmap(samples.sep.summer, sel_genes_summer, "Figures/Transcriptomics_Torgeir/heatmap_selected_genes_summer.pdf")
```


